### DISCUSSION TITLE
How should the custom mapping function look like for mapping a key onto two different cmp.functions?

### DISCUSSION THREAD

**Discussion by znxuz** [hrsh7th/nvim-cmp#1707] [Q&A] [answered]
For example if `cmp.visible()`, `<tab>` should map to `cmp.mapping.confirm({ select = true})`, or if it's not visible then it should invoke the completion menu via `cmp.mapping.complete()`. The function below somehow doesn't work:

```lua
function(_)
  if cmp.visible() then
    cmp.mapping.confirm({ select = true })
  else
    cmp.mapping.complete()
  end
end
```

I don't have any other existing mapping for `<tab>`.

**Comment by fitrh** [comment, marked as answer]
See `:h cmp-mapping`, at the end of section, it says

> Built-in mapping helpers are only available as a configuration option.
If you want to call nvim-cmp features directly, please use |cmp-function| instead.

It means that `cmp.mapping` is used in the form of `key = cmp.mapping.*`, using custom callback means you have to write it in the followinhg form

```lua
function(_)
  if cmp.visible() then
    cmp.confirm({ select = true })
  else
    cmp.complete()
  end
end
```

**Reply by znxuz** [reply]
Rookie mistake, thanks 👍🏼
